### PR TITLE
DATACOUCH-415 - Improved instantiation performance

### PR DESCRIPTION
**Pull request by ignacio-ifm**
Improved CouchbaseDocument and CouchbaseList instantiation performance

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@ignacio-ifm Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-couchbase&pullRequestId=176)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-data-couchbase&pullRequestId=176) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@ignacio-ifm Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-couchbase&pullRequestId=176)!

File: src/main/java/org/springframework/data/couchbase/core/mapping/CouchbaseDocumentSimpleTypes.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package org.springframework.data.couchbase.core.mapping;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.springframework.data.mapping.model.SimpleTypeHolder;
+
+public class CouchbaseDocumentSimpleTypes {
```
Comment (subalakr) [inline src/main/java/org/springframework/data/couchbase/core/mapping/CouchbaseDocumentSimpleTypes.java]: Rename the class to CouchbaseSimpleTypes as it doesn't seem specific CouchbaseDocument?

File: src/main/java/org/springframework/data/couchbase/core/mapping/CouchbaseDocumentSimpleTypes.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  static {
+    Set<Class<?>> simpleTypes = new HashSet<Class<?>>();
+    simpleTypes.add(CouchbaseDocument.class);
+    simpleTypes.add(CouchbaseList.class);
+    COUCHBASE_SIMPLE_TYPES = Collections.unmodifiableSet(simpleTypes);
+  }
+
+  private static final Set<Class<?>> COUCHBASE_SIMPLE_TYPES;
```
Comment (subalakr) [inline src/main/java/org/springframework/data/couchbase/core/mapping/CouchbaseDocumentSimpleTypes.java]: Can this be refactored with direct initializer using stream or anonymous class?

File: src/main/java/org/springframework/data/couchbase/core/mapping/CouchbaseDocumentSimpleTypes.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  public static final SimpleTypeHolder HOLDER = new SimpleTypeHolder(COUCHBASE_SIMPLE_TYPES, true);
```
Comment (subalakr) [inline src/main/java/org/springframework/data/couchbase/core/mapping/CouchbaseDocumentSimpleTypes.java]: Rename as DEFAULT similar to SimpleTypeHolder

File: src/main/java/org/springframework/data/couchbase/core/mapping/CouchbaseDocumentSimpleTypes.java (hunk 1 above)
Comment (ignacio-ifm) [reply src/main/java/org/springframework/data/couchbase/core/mapping/CouchbaseDocumentSimpleTypes.java]: Done.

File: src/main/java/org/springframework/data/couchbase/core/mapping/CouchbaseDocumentSimpleTypes.java (hunk 2 above)
Comment (ignacio-ifm) [reply src/main/java/org/springframework/data/couchbase/core/mapping/CouchbaseDocumentSimpleTypes.java]: Done.

File: src/main/java/org/springframework/data/couchbase/core/mapping/CouchbaseDocumentSimpleTypes.java (hunk 3 above)
Comment (ignacio-ifm) [reply src/main/java/org/springframework/data/couchbase/core/mapping/CouchbaseDocumentSimpleTypes.java]: Done.

File: src/main/java/org/springframework/data/couchbase/core/mapping/CouchbaseSimpleTypes.java [hunk 4]
```diff
@@ -28,16 +28,13 @@
...
 public abstract class CouchbaseSimpleTypes {
 
-  static {
-    Set<Class<?>> simpleTypes = new HashSet<Class<?>>();
-    simpleTypes.add(RawJsonDocument.class);
-    simpleTypes.add(JsonArray.class);
-    simpleTypes.add(Number.class);
-    COUCHBASE_SIMPLE_TYPES = Collections.unmodifiableSet(simpleTypes);
-  }
+  private static final Set<Class<?>> COUCHBASE_SIMPLE_TYPES = Stream.of(
```
Comment (subalakr) [inline src/main/java/org/springframework/data/couchbase/core/mapping/CouchbaseSimpleTypes.java]: This changes the functionality, the simple types added before were 
static {
    Set<Class<?>> simpleTypes = new HashSet<Class<?>>();
    simpleTypes.add(CouchbaseDocument.class);
    simpleTypes.add(CouchbaseList.class);
    COUCHBASE_SIMPLE_TYPES = Collections.unmodifiableSet(simpleTypes);
  }

**Comment by ignacio-ifm** [comment]
I have reset the PR to the initial state. Last change was a bit confusing. The point is CouchbaseSimpleTypes already exists as a class in the codebase with diferent types. I just followed the same pattern and naming of that class just to have a SimpleTypeHolder for the CouchbaseDocument class constructor (hence the name CouchbaseDocumentSimpleTypes).

Other way would be to add the static member to CouchbaseSimpleTypes (already existing) and name it somehow different than DEFAULT.

What do you think?

**Comment by subalakr** [comment]
Yes, the second option would be great, with the sets maybe named as simple types and additional types for holding Json and Couchbase types respectively.

**Comment by subalakr** [comment]
Merged

**Pull request closed without merging**